akka-stream/src/main/scala/akka/stream/impl/TraversalBuilder.scala

```diff
@@ -0,0 +1,985 @@
+ *    every output port should be assigned to (base + outToSlots(out.id)) where base is the number of the first
+ *    input port of the module (or the last unused input number if it has no input ports) and outToSlots is the
+ *    array provided by the traversal step.
+ *
+ * Since the above two rules always require local only computations (except a counter) this achieves
+ * positional independence of materializations.
+ *
+ * See the `TraversalTestUtils` class and the `testMaterialize` method for a simple example.
+ */
+final case class MaterializeAtomic(module: AtomicModule[Shape, Any], outToSlots: Array[Int]) extends Traversal {
+  override def toString: String = s"MaterializeAtomic($module, ${outToSlots.mkString("[", ", ", "]")})"
+
+  override def rewireFirstTo(relativeOffset: Int): Traversal = copy(outToSlots = Array(relativeOffset))
+}
+
+/**
+ * Traversal with no steps.
+ */
+object EmptyTraversal extends Traversal {
+  override def concat(that: Traversal): Traversal = that
+}
+
+sealed trait MaterializedValueOp extends Traversal
+
+case object Pop extends MaterializedValueOp
+case object PushNotUsed extends MaterializedValueOp
+final case class Transform(mapper: Any ⇒ Any) extends MaterializedValueOp
+final case class Compose(composer: (Any, Any) ⇒ Any) extends MaterializedValueOp
+
+final case class PushAttributes(attributes: Attributes) extends Traversal
+final case object PopAttributes extends Traversal
+
+final case class EnterIsland(islandTag: IslandTag, island: Traversal) extends Traversal {
+  override def rewireFirstTo(relativeOffset: Int): Traversal =
+    copy(island = island.rewireFirstTo(relativeOffset))
+}
+
+// Never embedded into actual traversal, used as a marker in AbsoluteTraversal
+final case class ExitIsland(islandGlobalOffset: Int, skippedSlots: Int, phase: PhaseIsland[Any]) extends Traversal
+
+object TraversalBuilder {
+
+  private val cachedEmptyCompleted = CompletedTraversalBuilder(PushNotUsed, 0, Map.empty, Attributes.none)
+
+  /**
+   * Assign ports their id, which is their position inside the Shape. This is used both by the GraphInterpreter
+   * and the layout system here.
+   */
+  private[impl] def initShape(shape: Shape): Unit = {
+    // Initialize port IDs
+    val inIter = shape.inlets.iterator
+    var i = 0
+    while (inIter.hasNext) {
+      inIter.next.id = i
+      i += 1
+    }
+
+    val outIter = shape.outlets.iterator
+    i = 0
+    while (outIter.hasNext) {
+      outIter.next.id = i
+      i += 1
+    }
+  }
+
+  def empty(attributes: Attributes = Attributes.none): TraversalBuilder = {
+    if (attributes eq Attributes.none) cachedEmptyCompleted
+    else CompletedTraversalBuilder(PushNotUsed, 0, Map.empty, attributes)
+  }
+
+  /**
+   * Create a generic traversal builder starting from an atomic module.
+   */
+  def atomic(module: AtomicModule[Shape, Any], attributes: Attributes = Attributes.none): TraversalBuilder = {
+    initShape(module.shape)
+
```

Comment:
Check if atomic is linear, if so, build a linear instead of a generic AtomicTraversalBuilder